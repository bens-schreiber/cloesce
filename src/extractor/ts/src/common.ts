export type Either<L, R> = { ok: false; value: L } | { ok: true; value: R };
export function left<L>(value: L): Either<L, never> {
  return { ok: false, value };
}
export function right<R>(value: R): Either<never, R> {
  return { ok: true, value };
}

export enum ExtractorErrorCode {
  UnknownType,
  MultipleGenericType,
  InvalidIncludeTree,
  UnknownNavigationPropertyReference,
  InvalidNavigationPropertyReference,
  MissingNavigationPropertyReference,
  MissingManyToManyUniqueId,
  MissingPrimaryKey,
  MissingWranglerEnv,
  TooManyWranglerEnvs,
  TooManyMiddlewares,
  MissingMiddlewareMethod,
}

const errorInfoMap: Record<
  ExtractorErrorCode,
  { description: string; suggestion: string }
> = {
  [ExtractorErrorCode.UnknownType]: {
    description: "Encountered an unknown or unsupported type",
    suggestion: "Refer to the documentation on valid Cloesce TS types",
  },
  [ExtractorErrorCode.MultipleGenericType]: {
    description: "Cloesce does not yet support types with multiple generics",
    suggestion:
      "Simplify your type to use only a single generic parameter, ie Foo<T>",
  },
  [ExtractorErrorCode.InvalidIncludeTree]: {
    description: "Invalid Include Tree",
    suggestion:
      "Include trees must only contain references to a model's navigation properties.",
  },
  [ExtractorErrorCode.UnknownNavigationPropertyReference]: {
    description: "Unknown Navigation Property Reference",
    suggestion:
      "Verify that the navigation property reference model exists, or create a model.",
  },
  [ExtractorErrorCode.InvalidNavigationPropertyReference]: {
    description: "Invalid Navigation Property Reference",
    suggestion: "Ensure the navigation property points to a valid model field",
  },
  [ExtractorErrorCode.MissingNavigationPropertyReference]: {
    description: "Missing Navigation Property Reference",
    suggestion:
      "Navigation properties require a foreign key model attribute reference",
  },
  [ExtractorErrorCode.MissingManyToManyUniqueId]: {
    description: "Missing unique id on Many to Many navigation property",
    suggestion:
      "Define a unique identifier field for the Many-to-Many relationship",
  },
  [ExtractorErrorCode.MissingPrimaryKey]: {
    description: "Missing primary key on a model",
    suggestion: "Add a primary key field to your model (e.g., `id: number`)",
  },
  [ExtractorErrorCode.MissingWranglerEnv]: {
    description: "Missing a wrangler environment definition in the project",
    suggestion: "Add a @WranglerEnv class in your project.",
  },
  [ExtractorErrorCode.TooManyWranglerEnvs]: {
    description: "Too many wrangler environments defined in the project",
    suggestion: "Consolidate or remove unused @WranglerEnv's",
  },
  [ExtractorErrorCode.TooManyMiddlewares]: {
    description: "Too many middleware classes defined in the project",
    suggestion: "Only one @Middleware class is allowed per project",
  },
  [ExtractorErrorCode.MissingMiddlewareMethod]: {
    description: "Middleware class must have a 'handle' method",
    suggestion: "Add a 'handle' method to your @Middleware class",
  },
};

export function getErrorInfo(code: ExtractorErrorCode) {
  return errorInfoMap[code];
}

export class ExtractorError {
  context?: string;
  snippet?: string;

  constructor(public code: ExtractorErrorCode) {}

  addContext(fn: (val: string | undefined) => string | undefined) {
    this.context = fn(this.context ?? "");
  }
}

export type HttpResult<T = unknown> = {
  ok: boolean;
  status: number;
  data?: T;
  message?: string;
};

export type CidlType =
  | "Void"
  | "Integer"
  | "Real"
  | "Text"
  | "Blob"
  | { Inject: string }
  | { Object: string }
  | { Nullable: CidlType }
  | { Array: CidlType }
  | { HttpResult: CidlType };

export function isNullableType(ty: CidlType): boolean {
  return typeof ty === "object" && ty !== null && "Nullable" in ty;
}

export enum HttpVerb {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE",
}

export interface NamedTypedValue {
  name: string;
  cidl_type: CidlType;
}

export interface ModelAttribute {
  value: NamedTypedValue;
  foreign_key_reference: string | null;
}

export interface ModelMethod {
  name: string;
  is_static: boolean;
  http_verb: HttpVerb;
  return_type: CidlType | null;
  parameters: NamedTypedValue[];
}

export type NavigationPropertyKind =
  | { OneToOne: { reference: string } }
  | { OneToMany: { reference: string } }
  | { ManyToMany: { unique_id: string } };

export interface NavigationProperty {
  var_name: string;
  model_name: string;
  kind: NavigationPropertyKind;
}

export function getNavigationPropertyCidlType(
  nav: NavigationProperty
): CidlType {
  return "OneToOne" in nav.kind
    ? { Object: nav.model_name }
    : { Array: { Object: nav.model_name } };
}

export interface Model {
  name: string;
  primary_key: NamedTypedValue;
  attributes: ModelAttribute[];
  navigation_properties: NavigationProperty[];
  methods: Record<string, ModelMethod>;
  data_sources: Record<string, DataSource>;
  source_path: string;
}

export interface PlainOldObject {
  name: string;
  attributes: NamedTypedValue[];
  source_path: string;
}

export interface CidlIncludeTree {
  [key: string]: CidlIncludeTree;
}

export interface DataSource {
  name: string;
  tree: CidlIncludeTree;
}

export interface WranglerEnv {
  name: string;
  source_path: string;
}

export interface MiddlewareMethod {
  name: string;
  parameters: NamedTypedValue[];
}

export interface Middleware {
  class_name: string;
  method: MiddlewareMethod;
  source_path: string;
}

export interface CloesceAst {
  version: string;
  project_name: string;
  language: "TypeScript";
  wrangler_env: WranglerEnv;
  models: Record<string, Model>;
  poos: Record<string, PlainOldObject>;
  middleware: Middleware | null;
}
