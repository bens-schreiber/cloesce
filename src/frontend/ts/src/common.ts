/// A partial type whose object keys may be partial as well
export type DeepPartial<T> = T extends (infer U)[]
  ? DeepPartial<U>[] // handle arrays specially
  : T extends object
    ? { [K in keyof T]?: DeepPartial<T[K]> }
    : T;

export type Either<L, R> = { ok: false; value: L } | { ok: true; value: R };
export function left<L>(value: L): Either<L, never> {
  return { ok: false, value };
}
export function right<R>(value: R): Either<never, R> {
  return { ok: true, value };
}

export enum ExtractorErrorCode {
  MissingExport,
  AppMissingDefaultExport,
  UnknownType,
  MultipleGenericType,
  DataSourceMissingStatic,
  InvalidPartialType,
  InvalidIncludeTree,
  UnknownNavigationPropertyReference,
  InvalidNavigationPropertyReference,
  MissingNavigationPropertyReference,
  MissingManyToManyUniqueId,
  MissingPrimaryKey,
  MissingWranglerEnv,
  TooManyWranglerEnvs,
  MissingFile,
}

const errorInfoMap: Record<
  ExtractorErrorCode,
  { description: string; suggestion: string }
> = {
  [ExtractorErrorCode.MissingExport]: {
    description: "All Cloesce types must be exported.",
    suggestion: "Add `export` to the class definition.",
  },
  [ExtractorErrorCode.AppMissingDefaultExport]: {
    description: "app.cloesce.ts does not export a CloesceApp by default",
    suggestion: "Export an instantiated CloesceApp in app.cloesce.ts",
  },
  [ExtractorErrorCode.UnknownType]: {
    description: "Encountered an unknown or unsupported type",
    suggestion: "Refer to the documentation on valid Cloesce TS types",
  },
  [ExtractorErrorCode.InvalidPartialType]: {
    description: "Partial types must only contain a model or plain old object",
    suggestion: "Refer to the documentation on valid Cloesce TS types",
  },
  [ExtractorErrorCode.MultipleGenericType]: {
    description: "Cloesce does not yet support types with multiple generics",
    suggestion:
      "Simplify your type to use only a single generic parameter, ie Foo<T>",
  },
  [ExtractorErrorCode.DataSourceMissingStatic]: {
    description: "Data Sources must be declared as static",
    suggestion: "Declare your data source as `static readonly`",
  },
  [ExtractorErrorCode.InvalidIncludeTree]: {
    description: "Invalid Include Tree",
    suggestion:
      "Include trees must only contain references to a model's navigation properties.",
  },
  [ExtractorErrorCode.UnknownNavigationPropertyReference]: {
    description: "Unknown Navigation Property Reference",
    suggestion:
      "Verify that the navigation property reference model exists, or create a model.",
  },
  [ExtractorErrorCode.InvalidNavigationPropertyReference]: {
    description: "Invalid Navigation Property Reference",
    suggestion: "Ensure the navigation property points to a valid model field",
  },
  [ExtractorErrorCode.MissingNavigationPropertyReference]: {
    description: "Missing Navigation Property Reference",
    suggestion:
      "Navigation properties require a foreign key model attribute reference",
  },
  [ExtractorErrorCode.MissingManyToManyUniqueId]: {
    description: "Missing unique id on Many to Many navigation property",
    suggestion:
      "Define a unique identifier field for the Many-to-Many relationship",
  },
  [ExtractorErrorCode.MissingPrimaryKey]: {
    description: "Missing primary key on a model",
    suggestion: "Add a primary key field to your model (e.g., `id: number`)",
  },
  [ExtractorErrorCode.MissingWranglerEnv]: {
    description: "Missing a wrangler environment definition in the project",
    suggestion: "Add a @WranglerEnv class in your project.",
  },
  [ExtractorErrorCode.TooManyWranglerEnvs]: {
    description: "Too many wrangler environments defined in the project",
    suggestion: "Consolidate or remove unused @WranglerEnv's",
  },
  [ExtractorErrorCode.MissingFile]: {
    description: "A specified input file could not be found",
    suggestion: "Verify the input file path is correct",
  },
};

export function getErrorInfo(code: ExtractorErrorCode) {
  return errorInfoMap[code];
}

export class ExtractorError {
  context?: string;
  snippet?: string;

  constructor(public code: ExtractorErrorCode) {}

  addContext(fn: (val: string | undefined) => string | undefined) {
    this.context = fn(this.context ?? "");
  }
}

export type HttpResult<T = unknown> = {
  ok: boolean;
  status: number;
  data?: T;
  message?: string;
};

/**
 * Dependency injection container, mapping an object type name to an instance of that object.
 *
 * Comes with the WranglerEnv and Request by default.
 */
export type InstanceRegistry = Map<string, any>;
export type MiddlewareFn = (
  request: Request,
  env: any,
  ir: InstanceRegistry,
) => Promise<HttpResult | undefined>;

export type KeysOfType<T, U> = {
  [K in keyof T]: T[K] extends U ? (K extends string ? K : never) : never;
}[keyof T];

/**
 * A container for middleware. If an instance is exported from `app.cloesce.ts`, it will be used in the
 * appropriate location, with global middleware happening before any routing occurs.
 */
export class CloesceApp {
  public global: MiddlewareFn[] = [];
  public model: Map<string, MiddlewareFn[]> = new Map();
  public method: Map<string, Map<string, MiddlewareFn[]>> = new Map();

  public useGlobal(m: MiddlewareFn) {
    this.global.push(m);
  }

  public useModel<T>(ctor: new () => T, m: MiddlewareFn) {
    if (this.model.has(ctor.name)) {
      this.model.get(ctor.name)!.push(m);
    } else {
      this.model.set(ctor.name, [m]);
    }
  }

  public useMethod<T>(
    ctor: new () => T,
    method: KeysOfType<T, (...args: any) => any>,
    m: MiddlewareFn,
  ) {
    if (!this.method.has(ctor.name)) {
      this.method.set(ctor.name, new Map());
    }

    const methods = this.method.get(ctor.name)!;
    if (!methods.has(method)) {
      methods.set(method, []);
    }

    methods.get(method)!.push(m);
  }
}

export type CrudKind = "POST" | "GET" | "LIST";

export type CidlType =
  | "Void"
  | "Integer"
  | "Real"
  | "Text"
  | "Blob"
  | "DataSource"
  | { Inject: string }
  | { Object: string }
  | { Partial: string }
  | { Nullable: CidlType }
  | { Array: CidlType }
  | { HttpResult: CidlType };

export function isNullableType(ty: CidlType): boolean {
  return typeof ty === "object" && ty !== null && "Nullable" in ty;
}

export enum HttpVerb {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE",
}

export interface NamedTypedValue {
  name: string;
  cidl_type: CidlType;
}

export interface ModelAttribute {
  value: NamedTypedValue;
  foreign_key_reference: string | null;
}

export interface ModelMethod {
  name: string;
  is_static: boolean;
  http_verb: HttpVerb;
  return_type: CidlType | null;
  parameters: NamedTypedValue[];
}

export type NavigationPropertyKind =
  | { OneToOne: { reference: string } }
  | { OneToMany: { reference: string } }
  | { ManyToMany: { unique_id: string } };

export interface NavigationProperty {
  var_name: string;
  model_name: string;
  kind: NavigationPropertyKind;
}

export function getNavigationPropertyCidlType(
  nav: NavigationProperty,
): CidlType {
  return "OneToOne" in nav.kind
    ? { Object: nav.model_name }
    : { Array: { Object: nav.model_name } };
}

export interface Model {
  name: string;
  primary_key: NamedTypedValue;
  attributes: ModelAttribute[];
  navigation_properties: NavigationProperty[];
  methods: Record<string, ModelMethod>;
  data_sources: Record<string, DataSource>;
  source_path: string;
}

export interface PlainOldObject {
  name: string;
  attributes: NamedTypedValue[];
  source_path: string;
}

export interface CidlIncludeTree {
  [key: string]: CidlIncludeTree;
}

export const NO_DATA_SOURCE = "none";
export interface DataSource {
  name: string;
  tree: CidlIncludeTree;
}

export interface WranglerEnv {
  name: string;
  source_path: string;
}

export interface CloesceAst {
  version: string;
  project_name: string;
  language: "TypeScript";
  wrangler_env: WranglerEnv;
  models: Record<string, Model>;
  poos: Record<string, PlainOldObject>;
  app_source: string | null;
}
