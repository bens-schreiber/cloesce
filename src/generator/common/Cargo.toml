[package]
name = "common"
version = "0.0.1"
edition = "2024"

[dependencies]
anyhow = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
serde_with = "3.14.1"
toml = { workspace = true }
