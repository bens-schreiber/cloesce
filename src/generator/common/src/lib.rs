pub mod builder;
pub mod err;

use std::collections::BTreeMap;
use std::path::PathBuf;

use err::GeneratorErrorKind;
use err::Result;
use serde::Deserialize;
use serde::Serialize;
use serde_with::MapPreventDuplicates;
use serde_with::serde_as;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum CidlType {
    /// No type
    Void,

    /// SQLite integer
    Integer,

    /// SQLite floating point number
    Real,

    /// SQLite string
    Text,

    /// SQLite large structured data
    Blob,

    /// A dependency injected instance, containing a type name.
    Inject(String),

    /// A model, or plain old object, containing the name of the class.
    Object(String),

    /// A part of a model or plain object, containing the name of the class.
    ///
    /// Only valid as a method argument.
    Partial(String),

    /// An array of any type
    Array(Box<CidlType>),

    /// A REST API response, which can contain any type or nothing.
    HttpResult(Box<CidlType>),

    /// A wrapper denoting the type within can be null.
    /// If the inner value is void, represents just null.
    Nullable(Box<CidlType>),
}

impl CidlType {
    /// Returns the inner part of an array if the type is an array
    pub fn unwrap_array(&self) -> Option<&CidlType> {
        match self {
            CidlType::Array(inner) => Some(inner),
            _ => None,
        }
    }

    /// Returns the root most CidlType, being any non Model/Array/Result.
    pub fn root_type(&self) -> &CidlType {
        match self {
            CidlType::Array(inner) => inner.root_type(),
            CidlType::HttpResult(inner) => inner.root_type(),
            CidlType::Nullable(inner) => inner.root_type(),
            t => t,
        }
    }

    pub fn is_nullable(&self) -> bool {
        matches!(self, CidlType::Nullable(_))
    }

    pub fn array(cidl_type: CidlType) -> CidlType {
        CidlType::Array(Box::new(cidl_type))
    }

    pub fn nullable(cidl_type: CidlType) -> CidlType {
        CidlType::Nullable(Box::new(cidl_type))
    }

    pub fn null() -> CidlType {
        CidlType::Nullable(Box::new(CidlType::Void))
    }
}

#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum HttpVerb {
    GET,
    POST,
    PUT,
    PATCH,
    DELETE,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash)]
pub struct NamedTypedValue {
    pub name: String,
    pub cidl_type: CidlType,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ModelAttribute {
    pub value: NamedTypedValue,
    pub foreign_key_reference: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ModelMethod {
    pub name: String,
    pub is_static: bool,
    pub http_verb: HttpVerb,
    pub return_type: CidlType,
    pub parameters: Vec<NamedTypedValue>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct IncludeTree(pub BTreeMap<String, IncludeTree>);

#[derive(Serialize, Deserialize, Debug)]
pub struct DataSource {
    pub name: String,
    pub tree: IncludeTree,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum NavigationPropertyKind {
    OneToOne { reference: String },
    OneToMany { reference: String },
    ManyToMany { unique_id: String },
}

#[derive(Serialize, Deserialize, Debug)]
pub struct NavigationProperty {
    pub var_name: String,
    pub model_name: String,
    pub kind: NavigationPropertyKind,
}

#[serde_as]
#[derive(Serialize, Deserialize, Debug)]
pub struct Model {
    pub name: String,
    pub primary_key: NamedTypedValue,
    pub attributes: Vec<ModelAttribute>,
    pub navigation_properties: Vec<NavigationProperty>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub methods: BTreeMap<String, ModelMethod>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub data_sources: BTreeMap<String, DataSource>,
    pub source_path: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PlainOldObject {
    pub name: String,
    pub attributes: Vec<NamedTypedValue>,
    pub source_path: PathBuf,
}

#[derive(Serialize, Deserialize)]
pub enum InputLanguage {
    TypeScript,
}

#[derive(Serialize, Deserialize)]
pub struct WranglerEnv {
    pub name: String,
    pub source_path: PathBuf,
}

#[serde_as]
#[derive(Serialize, Deserialize)]
pub struct CloesceAst {
    pub version: String,
    pub project_name: String,
    pub language: InputLanguage,
    pub wrangler_env: WranglerEnv,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub models: BTreeMap<String, Model>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub poos: BTreeMap<String, PlainOldObject>,
}

impl CloesceAst {
    pub fn from_json(path: &std::path::Path) -> Result<CloesceAst> {
        let cidl_contents = std::fs::read_to_string(path).map_err(|e| {
            GeneratorErrorKind::InvalidInputFile
                .to_error()
                .with_context(e.to_string())
        })?;
        serde_json::from_str::<CloesceAst>(&cidl_contents).map_err(|e| {
            GeneratorErrorKind::InvalidInputFile
                .to_error()
                .with_context(e.to_string())
        })
    }

    fn is_valid_object_ref(&self, o: &str) -> bool {
        self.models.contains_key(o) || self.poos.contains_key(o)
    }

    /// Ensures all `CidlTypes` are logically correct for the area, essentially doing
    /// the first level of semantic analysis for the generator.
    ///
    /// Returns error on
    /// - Model attributes with invalid SQL types
    /// - Primary keys with invalid SQL types
    /// - Invalid Model or Method map K/V
    /// - Unknown navigation property model references
    /// - Unknown model references in method parameters
    /// - Invalid method parameter types
    pub fn validate_types(&self) -> Result<()> {
        let ensure_valid_sql_type = |model: &Model, value: &NamedTypedValue| {
            let inner = match &value.cidl_type {
                CidlType::Nullable(inner) if matches!(inner.as_ref(), CidlType::Void) => {
                    fail!(GeneratorErrorKind::NullSqlType)
                }
                CidlType::Nullable(inner) => inner.as_ref(),
                other => other,
            };

            ensure!(
                matches!(
                    inner,
                    CidlType::Integer | CidlType::Real | CidlType::Text | CidlType::Blob
                ),
                GeneratorErrorKind::InvalidSqlType,
                "{}.{}",
                model.name,
                value.name
            );

            Ok(())
        };

        for (model_name, model) in &self.models {
            // Validate record
            ensure!(
                *model_name == model.name,
                GeneratorErrorKind::InvalidMapping,
                "Model record key did not match it's model name? {} : {}",
                model_name,
                model.name
            );

            // Validate PK
            ensure_valid_sql_type(model, &model.primary_key)?;

            // Validate attributes
            for a in &model.attributes {
                ensure_valid_sql_type(model, &a.value)?;

                if let Some(fk_model) = &a.foreign_key_reference {
                    // Validate the fk's model exists
                    ensure!(
                        self.models.contains_key(fk_model.as_str()),
                        GeneratorErrorKind::UnknownObject,
                        "{}.{} => {}?",
                        model.name,
                        a.value.name,
                        fk_model
                    );
                }
            }

            // Validate navigation props
            for nav in &model.navigation_properties {
                ensure!(
                    self.models.contains_key(nav.model_name.as_str()),
                    GeneratorErrorKind::UnknownObject,
                    "{} => {}?",
                    model.name,
                    nav.model_name
                );
            }

            // Validate methods
            for (method_name, method) in &model.methods {
                // Validate record
                ensure!(
                    *method_name == method.name,
                    GeneratorErrorKind::InvalidMapping,
                    "Method record key did not match it's method name? {}: {}",
                    method_name,
                    method.name
                );

                // Validate return type
                match &method.return_type {
                    CidlType::Object(o) => {
                        ensure!(
                            self.is_valid_object_ref(o),
                            GeneratorErrorKind::UnknownObject,
                            "{}.{}",
                            model.name,
                            method.name
                        );
                    }
                    CidlType::Partial(_) => {
                        fail!(
                            GeneratorErrorKind::UnexpectedPartialReturn,
                            "{}.{}",
                            model.name,
                            method.name,
                        )
                    }
                    _ => {}
                }

                // Validate method params
                for param in &method.parameters {
                    let root_type = param.cidl_type.root_type();

                    match root_type {
                        CidlType::Void => {
                            fail!(
                                GeneratorErrorKind::UnexpectedVoid,
                                "{}.{}.{}",
                                model.name,
                                method.name,
                                param.name
                            )
                        }
                        CidlType::Object(o) | CidlType::Partial(o) => {
                            ensure!(
                                self.is_valid_object_ref(o),
                                GeneratorErrorKind::UnknownObject,
                                "{}.{}.{}",
                                model.name,
                                method.name,
                                param.name
                            );

                            // TODO: remove this
                            if method.http_verb == HttpVerb::GET {
                                fail!(
                                    GeneratorErrorKind::NotYetSupported,
                                    "GET Requests currently do not support model parameters {}.{}.{}",
                                    model.name,
                                    method.name,
                                    param.name
                                )
                            }
                        }
                        _ => {
                            // Ignore
                        }
                    }
                }
            }
        }

        Ok(())
    }
}
