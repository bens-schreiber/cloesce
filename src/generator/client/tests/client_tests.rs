use anyhow::Result;
use common::CloesceAst;
use insta::assert_snapshot;

use std::path::PathBuf;

#[test]
fn test_generate_client_snapshot() -> Result<()> {
    // Arrange
    let ast = {
        let cidl_path = PathBuf::from("../../test_fixtures/cidl.json");
        let cidl_contents = std::fs::read_to_string(cidl_path)?;
        serde_json::from_str::<CloesceAst>(&cidl_contents)?
    };

    // Act
    let client = client::generate_client_api(ast, "http://localhost:1000/api".to_string());

    // Assert
    assert_snapshot!("generated_client", client);
    Ok(())
}
