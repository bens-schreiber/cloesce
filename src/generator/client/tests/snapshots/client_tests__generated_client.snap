---
source: client/tests/client_tests.rs
expression: client
---
type Ok<T> = { ok: true; data: T };
type Err = { ok: false; status: number; message: string };
type Result<T> = Ok<T> | Err;

function instantiateModelArray<T>(data: any, ctor: { new(): T }): T[] {
  if (Array.isArray(data)) {
    return data.map(x => instantiateModelArray(x, ctor)).flat();
  }
  return [Object.assign(new ctor(), data)];
}

export class ExampleModel {
  id: number;
  rating: number;
  description: string | null;
  data_blob: Uint8Array;

  static async create_example(
        payload: ExampleModel
  ): Promise<Result<ExampleModel>> {
    const url = `http://localhost:1000/api/ExampleModel/create_example`;
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
            payload
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    const instantiated = Object.assign(new ExampleModel(), data);
    return { ok: true, data: instantiated };
  }

  static async read_example(
        id: number | null
  ): Promise<Result<string[]>> {
    const url = `http://localhost:1000/api/ExampleModel/read_example`;
    const res = await fetch(url, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
            id
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    return { ok: true, data };
  }

  static async list_examples(
  ): Promise<Result<ExampleModel[]>> {
    const url = `http://localhost:1000/api/ExampleModel/list_examples`;
    const res = await fetch(url, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    const instantiated = instantiateModelArray(data, ExampleModel);
    return { ok: true, data: instantiated };
  }

  static async update_example(
        id: number, 
        rating: number | null
  ): Promise<Result<number>> {
    const url = `http://localhost:1000/api/ExampleModel/update_example`;
    const res = await fetch(url, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
            id, 
            rating
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    return { ok: true, data };
  }

  static async patch_example(
        description: string | null
  ): Promise<Result<number>> {
    const url = `http://localhost:1000/api/ExampleModel/patch_example`;
    const res = await fetch(url, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
            description
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    return { ok: true, data };
  }

  static async delete_example(
        id: number
  ): Promise<Result<Uint8Array>> {
    const url = `http://localhost:1000/api/ExampleModel/delete_example`;
    const res = await fetch(url, {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
            id
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    return { ok: true, data };
  }

  async get_description(
  ): Promise<Result<string>> {
    const url = `http://localhost:1000/api/ExampleModel/${this.id}/get_description`;
    const res = await fetch(url, {
      method: "GET",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    return { ok: true, data };
  }

  async increment_rating(
        amount: number
  ): Promise<Result<number>> {
    const url = `http://localhost:1000/api/ExampleModel/${this.id}/increment_rating`;
    const res = await fetch(url, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
            amount
      })
    });
    if (!res.ok) {
      const message = await res.text();
      return {
        ok: false,
        status: res.status,
        message
      };
    }

    const data = await res.json();
    return { ok: true, data };
  }


}
